### PR TITLE
Fix Date from CSV parsing

### LINKED ISSUE #39248: Parsing out of range dates in CSV lead to unexpected result
When you parse Date column which have values bigger than by different format 

```sql
DROP TABLE IF EXISTS test_date_imports;
create temporary table test_date_imports (f String, t Date);
insert into test_date_imports(t) format TSV 2200-01-01
insert into test_date_imports format CSV 'csv',2200-01-01
insert into test_date_imports format JSONEachRow {"f":"JSONEachRow","t":"2200-01-01"}
insert into test_date_imports values ('values','2200-01-01')

SELECT * FROM test_date_imports FORMAT PrettyCompactMonoBlock;

┌─f───────────┬──────────t─┐
│             │ 2149-06-06 │
│ csv         │ 2020-07-27 │<-- it looks like parsing happened via Date32 or DateTime64 path, and then casted with overflow
│ JSONEachRow │ 2149-06-06 │
│ values      │ 2149-06-06 │
└─────────────┴────────────┘

TRUNCATE TABLE test_date_imports;
insert into test_date_imports(t) format TSV 9999-01-01
insert into test_date_imports format CSV 'csv',9999-01-01
insert into test_date_imports format JSONEachRow {"f":"JSONEachRow","t":"9999-01-01"}
insert into test_date_imports values ('values','9999-01-01')

SELECT * FROM test_date_imports FORMAT PrettyCompactMonoBlock

SELECT * FROM test_date_imports FORMAT PrettyCompactMonoBlock
┌─f───────────┬──────────t─┐
│             │ 2149-06-06 │
│ csv         │ 2104-06-06 │
│ JSONEachRow │ 2149-06-06 │
│ values      │ 2149-06-06 │
└─────────────┴────────────┘

4 rows in set. Elapsed: 0.002 sec. 
```

P.S. It's not the edge case found by QA. It's a real-life case from the end-user report. IRL it was the '[end of time](https://softwareengineering.stackexchange.com/questions/164843/is-there-a-constant-for-end-of-time)' date exported by another database as '9999-12-31' which was always stored in ClickHouse as '2149-06-06' but after importing from CSV as '2104-06-06'

### PR DESCRIPTION
**Pull request by zvonand**
Fix #39248

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix parsing out-of-range Date from CSV:

### CODE REVIEW

**Comment by evillique** [comment]
Could you add a stateless test with the example from the issue?

File: src/DataTypes/Serializations/SerializationDate.cpp [hunk 1]
```diff
@@ -76,9 +76,22 @@ void SerializationDate::serializeTextCSV(const IColumn & column, size_t row_num,
...
-    readCSV(value, istr);
-    assert_cast<ColumnUInt16 &>(column).getData().push_back(value.getDayNum());
+    DayNum x;
+
+    if (istr.eof())
+        throwReadAfterEOF();
+
+    char maybe_quote = *istr.position();
+
+    if (maybe_quote == '\'' || maybe_quote == '\"')
```
Comment (filimonov) [inline src/DataTypes/Serializations/SerializationDate.cpp:86]: There is a setting with name like 'Input format double quotes'

Comment (zvonand) [reply src/DataTypes/Serializations/SerializationDate.cpp:86]: you mean 'allow_single_quotes' and 'allow_double_quotes'? They were not used here before. They used only in readCSVString

**Comment by zvonand** [comment]
AST looks unrelated to me

**Comment by alexey-milovidov** [comment]
> Fix Date parsing from CSV

The changelog entry is not descriptive enough.

**Review by evillique — COMMENTED** [review]
It seems like a more elegant solution would be to add the following:
Around src/IO/ReadHelpers.h:1178
```
inline void readCSV(DayNum & x, ReadBuffer & buf) { readCSVSimple(x, buf); }
```
Around src/IO/ReadHelpers.h:1097
```
inline void readText(DayNum & x, ReadBuffer & buf) { readDateText(x, buf); }
```
And to only change src/DataTypes/Serializations/SerializationDate.cpp:79 to
```
DayNum value;
readCSV(value, istr);
assert_cast<ColumnUInt16 &>(column).getData().push_back(value);
```

**Comment by zvonand** [comment]
@evillique updated with your recommendation

**Review by evillique — APPROVED** [review]

**Comment by evillique** [comment]
AST fuzzer (tsan): https://github.com/ClickHouse/ClickHouse/issues/41848

**Pull request merged**